src/github.com/couchbase/sync_gateway/base/sharded_sequence_clock.go

```diff
@@ -0,0 +1,294 @@
+	shardedClockExpvars = expvar.NewMap("syncGateway_index_clocks")
+}
+
+const (
+	kIndexPartitionKey       = "_idxPartitionMap"
+	kIndexPrefix             = "_idx"
+	kCountKeyFormat          = "_idx_c:%s:count"    // key
+	kClockPartitionKeyFormat = "_idx_c:%s:clock-%d" // key, partition index
+)
+
+type PartitionStorage struct {
+	Uuid  string   `json:"uuid"`
+	Index uint16   `json:"index"`
+	VbNos []uint16 `json:"vbNos"`
+}
+
+type IndexPartitionMap map[uint16]uint16 // Maps vbuckets to index partition value
+
+type IndexPartitions struct {
+	VbMap         IndexPartitionMap
+	PartitionDefs []PartitionStorage
+}
+
+func NewIndexPartitions(partitions []PartitionStorage) *IndexPartitions {
+
+	indexPartitions := &IndexPartitions{
+		PartitionDefs: make([]PartitionStorage, len(partitions)),
+	}
+	copy(indexPartitions.PartitionDefs, partitions)
+
+	indexPartitions.VbMap = make(map[uint16]uint16)
+	for _, partition := range partitions {
+		for _, vbNo := range partition.VbNos {
+			indexPartitions.VbMap[vbNo] = partition.Index
+		}
+	}
+	return indexPartitions
+}
+
+// Index partitionsfor testing
+func TestIndexPartitions() *IndexPartitions {
+
+	maxVbNo := uint16(1024)
+	partitions := make(IndexPartitionMap, maxVbNo)
+
+	numPartitions := uint16(32)
+	partitionDefs := make([]PartitionStorage, numPartitions)
+	vbPerPartition := maxVbNo / numPartitions
+	for partition := uint16(0); partition < numPartitions; partition++ {
+		storage := PartitionStorage{
+			Index: partition,
+			VbNos: make([]uint16, vbPerPartition),
+		}
+		for index := uint16(0); index < vbPerPartition; index++ {
+			vb := partition*vbPerPartition + index
+			partitions[vb] = partition
+			storage.VbNos = append(storage.VbNos, vb)
+		}
+		partitionDefs[partition] = storage
+	}
+	return &IndexPartitions{
+		VbMap:         partitions,
+		PartitionDefs: partitionDefs,
+	}
+}
+
+// Index partition map for testing
+func TestIndexPartitionMap() IndexPartitionMap {
+	return TestIndexPartitions().VbMap
+}
+
+// ShardedClock maintains the collection of clock shards (ShardedClockPartitions), and also manages
+// the counter for the clock.
+type ShardedClock struct {
+	baseKey       string                            // key prefix used to build keys for clock component docs
+	counter       uint64                            // count value for clock
+	countKey      string                            // key used to incr count value
+	partitionMap  *IndexPartitions                  // Index partition map
+	partitions    map[uint16]*ShardedClockPartition // Clock partitions - one doc written per partition
+	bucket        Bucket                            // Bucket used to store clocks
+	partitionKeys []string                          // Keys for all partitions.  Convenience to avoid rebuilding set from partitions
+}
+
+// ShardedClockPartition manages storage for one clock partition, where a clock partition is a set of
+// {vb, seq} values for a subset of vbuckets.
+type ShardedClockPartition struct {
```

Comment:
The overlapping terminology of `partition` vs `shard` is a bit confusing.  In CBGT, I believe `partition` is essentially a more generic term for `vbucket` (eg, [MaxPartitionsPerPIndex](https://github.com/couchbaselabs/cbgt/blob/a1de7075393c19c5d7f99e64f8e6aece567926a8/defs.go#L58) is essentially saying MaxVbucketsPerPindex).  

In the way partition is being used here, it sounds like partition is being used more like as a synonym for "pindex", right?  

I'm just worried that the terminology delta will make things more confusing.  I wonder if it would make more sense to call this a "Shard" instead of a "Partition".

I don't think this should block the PR, but just putting it down as food for thought for a future discussion.